src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -0,0 +1,434 @@
+            "shard " << std::to_string(index_) <<
+            " ledger seq " << std::to_string(l->info().seq) <<
+            " already stored";
+        return false;
+    }
+    if (boost::icl::length(storedSeqs_) >=
+        (index_ == detail::genesisShardIndex ?
+            detail::genesisNumLedgers : ledgersPerShard) - 1)
+    {
+        if (backend_->fdlimit() != 0)
+        {
+            remove(control_);
+            updateFileSize();
+        }
+        complete_ = true;
+        storedSeqs_.clear();
+
+        JLOG(j_.debug()) <<
+            "shard " << std::to_string(index_) << " complete";
+    }
+    else
+    {
+        storedSeqs_.insert(l->info().seq);
+        lastStored_ = l;
+        if (backend_->fdlimit() != 0 && !saveControl())
+            return false;
+    }
+
+    JLOG(j_.debug()) <<
+        "shard " << std::to_string(index_) <<
+        " ledger seq " << std::to_string(l->info().seq) <<
+        " stored";
+
+    return true;
+}
+
+boost::optional<std::uint32_t>
+Shard::prepare()
+{
+    if (storedSeqs_.empty())
+         return lastSeq_;
+    return prevMissing(storedSeqs_, 1 + lastSeq_, firstSeq_);
+}
+
+bool
+Shard::hasLedger(std::uint32_t seq) const
+{
+    if (seq < firstSeq_ || seq > lastSeq_)
+        return false;
+    if (complete_)
+        return true;
+    return boost::icl::contains(storedSeqs_, seq);
+}
+
+void
+Shard::validate(Application& app)
+{
+    uint256 hash;
+    std::uint32_t seq;
+    std::shared_ptr<Ledger> l;
+    // Find the hash of the last ledger in this shard
+    {
+        std::tie(l, seq, hash) = loadLedgerHelper(
+            "WHERE LedgerSeq >= " + std::to_string(lastSeq_) +
+            " order by LedgerSeq desc limit 1", app);
+        if (!l)
+        {
+            JLOG(j_.fatal()) <<
+                "shard " << std::to_string(index_) <<
+                " unable to validate. No lookup data";
+            return;
+        }
+        if (seq != lastSeq_)
+        {
+            l->setImmutable(app.config());
+            boost::optional<uint256> h;
+            try
+            {
+                h = hashOfSeq(*l, lastSeq_, j_);
+            }
+            catch (std::exception const& e)
+            {
+                JLOG(j_.fatal()) <<
+                    "exception: " << e.what();
+                return;
+            }
+            if (!h)
+            {
+                JLOG(j_.fatal()) <<
+                    "shard " << std::to_string(index_) <<
+                    " No hash for last ledger seq " <<
+                    std::to_string(lastSeq_);
+                return;
+            }
+            hash = *h;
+            seq = lastSeq_;
+        }
+    }
+
+    JLOG(j_.fatal()) <<
+        "Validating shard " << std::to_string(index_) <<
+        " ledgers " << std::to_string(firstSeq_) << "-" <<
+        std::to_string(lastSeq_);
+
+    // Use a short age to keep memory consumption low
+    TaggedCache<uint256, NodeObject>::clock_type::rep const savedAge {
```

Comment:
The shard object returns a ref to its caches used by the database from various operations. In order to use a local cache you would need to swap it out with the class member. Because validate should only be called during the `application` setup, the caches must be empty before and after validate, so there would be no benefit.